modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -40,216 +41,359 @@
+    }
+
+    /**
+     * Copy the version from a search hit to the action request.
+     */
+    protected void copyVersion(ActionRequest<?> request, long version) {
+        if (request instanceof IndexRequest) {
+            ((IndexRequest) request).version(version);
+        } else if (request instanceof DeleteRequest) {
+            ((DeleteRequest) request).version(version);
+        } else {
+            throw new UnsupportedOperationException("unable to copy [version] to action request [" + request.getClass() + "]");
         }
+    }
 
-        /*
-         * It'd be lovely to only set the source if we know its been modified
-         * but it isn't worth keeping two copies of it around just to check!
-         */
-        index.source((Map<String, Object>) resultCtx.remove(SourceFieldMapper.NAME));
 
-        Object newValue = ctx.remove(IndexFieldMapper.NAME);
-        if (false == doc.index().equals(newValue)) {
-            scriptChangedIndex(index, newValue);
-        }
-        newValue = ctx.remove(TypeFieldMapper.NAME);
-        if (false == doc.type().equals(newValue)) {
-            scriptChangedType(index, newValue);
-        }
-        newValue = ctx.remove(IdFieldMapper.NAME);
-        if (false == doc.id().equals(newValue)) {
-            scriptChangedId(index, newValue);
+    /**
+     * Copy the version type from a search hit to the action request.
+     */
+    protected void copyVersionType(ActionRequest<?> request, VersionType versionType) {
+        if (request instanceof IndexRequest) {
+            ((IndexRequest) request).versionType(versionType);
+        } else if (request instanceof DeleteRequest) {
+            ((DeleteRequest) request).versionType(versionType);
+        } else {
+            throw new UnsupportedOperationException("unable to copy [version_type] to action request [" + request.getClass() + "]");
         }
-        newValue = ctx.remove(VersionFieldMapper.NAME);
-        if (false == Objects.equals(oldVersion, newValue)) {
-            scriptChangedVersion(index, newValue);
+    }
+
+    /**
+     * Copy the parent from a search hit to the action request.
+     */
+    protected void copyParent(ActionRequest<?> request, String parent) {
+        if (request instanceof IndexRequest) {
+            ((IndexRequest) request).parent(parent);
+        } else if (request instanceof DeleteRequest) {
+            ((DeleteRequest) request).parent(parent);
+        } else {
+            throw new UnsupportedOperationException("unable to copy [parent] to action request [" + request.getClass() + "]");
         }
-        newValue = ctx.remove(ParentFieldMapper.NAME);
-        if (false == Objects.equals(oldParent, newValue)) {
-            scriptChangedParent(index, newValue);
+    }
+
+    /**
+     * Copy the routing from a search hit to the action request.
+     */
+    protected void copyRouting(ActionRequest<?> request, String routing) {
+        if (request instanceof IndexRequest) {
```

Comment:
Honestly maybe wrapping the requests in another object - there is one interface and two implementations and they handle either arm of the instanceof check?